### DISCUSSION TITLE
考虑 meta StyleX 代替传统的 css-in-js

### DISCUSSION THREAD

**Discussion by guaijie** [ant-design/ant-design#46449] [General]
### What problem does this feature solve?

虽然 css-in-js 的方案能轻松实现主题切换，但这意味着自定义的组件也需要借助 css-in-js, 因为我们无法在样式文件中（如： less, css ）等应用 token（当然也可以借助 hook 和 style，但。。。）。而使用 css-in-js 总有无法避免的运行时开销。特别是开发者在使用时，很难做好这些处理。meta StyleX 的开源方案似乎给出了一种较好解决方式。但并非 StyleX 就意味着完美的解决方案，或许能从其中给出不错的启示

### What does the proposed API look like?

no response

**Comment by arvinxx** [comment]
less 做主题已经过时了，建议放弃吧。

css/sass 要用的话也可以。antd 现在也有 cssvar 模式了，理论上用 cssvar 也能兼容。

> 而使用 css-in-js 总有无法避免的运行时开销

抛开剂量谈毒性没意义。不然我还说 react 也有一层运行时开销呢，咋难道你还因此回到 jquery 手撸 dom 的时候么？

就事论事的话，建议给出你在实际业务中，在现有 css-in-js 模式下，明显是由于 cssinjs 造成的渲染瓶颈的案例来讨论。

就我自己这一年的使用体感来说，cssinjs 造成的性能开销远没有你用 less / sass 但是没做性能优化时的 rerender 开销大。

> meta StyleX 的开源方案似乎给出了一种较好解决方式

styleX 本质上也是一种零运行时的 cssinjs 方案，但它是否适用于 antd 并不一定。

**Reply by guaijie** [reply]
考虑运行时开销都在程序打包编译后。开发过程的是难以避免的。

**Reply by zombieJ** [reply]
参考 Blog，antd 是即可以运行时生成样式、又可以在编译阶段生成样式：
* 静态编译：https://ant-design.antgroup.com/docs/blog/extract-ssr-cn
* 混合态样式水合：https://ant-design.antgroup.com/docs/blog/hydrate-cssinjs-cn
* 使用 css var 实现动态主题能力：https://ant-design.antgroup.com/docs/blog/css-var-plan-cn

**Comment by yoyo837** [comment]
应用级的css-in-js和库的css-in-js还是有点区别吧
